### PR TITLE
Add 'otx test' command

### PR DESCRIPTION
**Pull request by harimkang**
### Summary

- Add the `checkpoint` attribute to TrainConfig
- Apply `ckpt_path=cfg.checkpoint` to `trainer.fit` in train function (Default=None)
- Implement test function
- Connect test to the draft CLI (`otx test`)
- Make the OTX CLI integration test an e2e test and add `otx test` to it
- Add segmentation dataset test subset into tests/asset

**The CLI is still in draft and needs further refinement later.

### How to test

```console
otx train +recipe=detection/yolox_tiny.yaml \
base.data_dir=tests/assets/car_tree_bug \
base.work_dir=otx_workspace/otx_train_yolox_tiny \
base.output_dir=otx_workspace/otx_train_yolox_tiny/outputs \
model.otx_model.config.bbox_head.num_classes=3

otx test +recipe=detection/yolox_tiny.yaml \
base.data_dir=tests/assets/car_tree_bug \
base.work_dir=otx_workspace/otx_test_yolox_tiny \
base.output_dir=otx_workspace/otx_test_yolox_tiny/outputs \
model.otx_model.config.bbox_head.num_classes=3 \
checkpoint=otx_workspace/otx_train_yolox_tiny/outputs/checkpoints/epoch_000.ckpt
```

### Checklist

- [ ] I have added unit tests to cover my changes.​
- [ ] I have added integration tests to cover my changes.​
- [ ] I have added e2e tests for validation.
- [ ] I have added the description of my changes into CHANGELOG in my target branch (e.g., [CHANGELOG](https://github.com/openvinotoolkit/training_extensions/blob/develop/CHANGELOG.md) in develop).​
- [ ] I have updated the documentation in my target branch accordingly (e.g., [documentation](https://github.com/openvinotoolkit/training_extensions/tree/develop/docs) in develop).
- [ ] I have [linked related issues](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

### License

- [ ] I submit _my code changes_ under the same [Apache License](https://github.com/openvinotoolkit/training_extensions/blob/develop/LICENSE) that covers the project.
      Feel free to contact the maintainers if that's a concern.
- [ ] I have updated the license header for each file (see an example below).

```python
# Copyright (C) 2023 Intel Corporation
# SPDX-License-Identifier: Apache-2.0
```

### CODE REVIEW

**Review by vinnamkim — DISMISSED** [review]
LGTM. However, would you please add this to `for_developers/setup_guide.md` for other developers to look at it easily from the codebase.
image

File: src/otx/core/engine/test.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+import hydra
+
+from otx.core.config import TrainConfig
+
+if TYPE_CHECKING:
+    from lightning import LightningModule, Trainer
+
+
+def test(cfg: TrainConfig) -> tuple[Trainer, dict[str, Any]]:
```
Comment (kprokofi) [inline src/otx/core/engine/test.py:18]: Should we move to class based Engine?

File: src/otx/core/engine/train.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def train(cfg: TrainConfig) -> tuple[Trainer, dict[str, Any]]:
 
     if cfg.train:
         log.info("Starting training!")
-        trainer.fit(model=model, datamodule=datamodule)
+        trainer.fit(model=model, datamodule=datamodule, ckpt_path=cfg.checkpoint)
```
Comment (kprokofi) [inline src/otx/core/engine/train.py:97]: what is ckpt_path? Is it for resume?

**Review by kprokofi — DISMISSED** [review]

**Review by sungmanc — DISMISSED** [review]

File: src/otx/core/engine/test.py (hunk 1 above)
Comment (harimkang) [reply src/otx/core/engine/test.py:18]: Engines should also be considered, but we don't use them at the moment. I'll revisit the Engine class in the future, along with the CLI.

File: src/otx/core/engine/train.py (hunk 2 above)
Comment (harimkang) [reply src/otx/core/engine/train.py:97]: Yes, that's correct.

**Comment by harimkang** [comment]
> LGTM. However, would you please add this to `for_developers/setup_guide.md` for other developers to look at it easily from the codebase. image

I added the `otx test` related commands to the `setup_guide.md`. https://github.com/openvinotoolkit/training_extensions/pull/2707/commits/6546e86d206b6d40e445ab746a19428d25cbd9c3

**Comment by harimkang** [comment]
There was also an issue with the instance of GlobalHydra not being cleaned up, so I modified it to run at the end of each cli function (train, test). https://github.com/openvinotoolkit/training_extensions/commit/6546e86d206b6d40e445ab746a19428d25cbd9c3

File: src/otx/cli/test.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    initialize(config_path="../config", version_base="1.3", job_name="otx_test")
+    cfg = compose(config_name="test", overrides=overrides, return_hydra_config=True)
+    configure_hydra_outputs(cfg)
+
+    # test the model
+    from otx.core.engine.test import test
+    metric_dict, _ = test(cfg)
+
+    # Clear the Hydra instances.
+    GlobalHydra().clear()
```
Comment (vinnamkim) [inline src/otx/cli/test.py]: `initialize` seems Python context manager. Please see https://github.com/facebookresearch/hydra/blob/2f7628756b953fa35d8f7a91fec1018a55cac43b/hydra/initialize.py#L98-L102.
Can we replace this with Python `with` statement? If possible, we must apply it to `train.py` as well.

Comment (harimkang) [reply src/otx/cli/test.py]: Thanks, I'll try

Comment (harimkang) [reply src/otx/cli/test.py]: This seems to work fine. Thanks. - https://github.com/openvinotoolkit/training_extensions/pull/2707/commits/c5ac6a30b96c5400a1c9f88d3f4d2ca5e50ea2e5

**Review by vinnamkim — APPROVED** [review]

**Review by sungmanc — APPROVED** [review]

**Review by jaegukhyun — APPROVED** [review]

**Pull request merged**